packages/aead/src/aad.rs

```diff
@@ -0,0 +1,173 @@
+use std::borrow::Cow;
+
+pub struct Aad<'a>(Cow<'a, [u8]>);
+
+impl<'a> Aad<'a> {
+    pub fn empty() -> Self {
+        Aad(Cow::Borrowed(&[]))
+    }
+
+    pub fn new_owned<I>(aad: I) -> Self
+    where
+        I: IntoIterator<Item = u8>,
+    {
+        // Collect the iterator into a Vec<u8>
+        let aad: Vec<u8> = aad.into_iter().collect();
+        // Convert the Vec<u8> into a Cow<[u8]>
+        // and wrap it in Aad
+        Aad(Cow::Owned(aad))
+    }
+
+    pub fn from_slice(slice: &'a [u8]) -> Self {
+        Aad(Cow::Borrowed(slice))
+    }
+
+    pub fn as_bytes(&self) -> &[u8] {
+        self.0.as_ref()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+
+    pub fn into_owned(self) -> Aad<'a> {
+        match self.0 {
+            x @ Cow::Borrowed(_) => Self(x.into_owned().into()),
+            Cow::Owned(_) => self,
+        }
+    }
+}
+
+impl<'a> Extend<u8> for Aad<'a> {
+    fn extend<T: IntoIterator<Item = u8>>(&mut self, iter: T) {
+        self.0.to_mut().extend(iter);
+    }
+}
+
+pub trait IntoAad<'a> {
+    fn into_aad(self) -> Aad<'a>
+    where
+        Self: Sized;
+}
```

Comment:
What's the reason behind this trait instead of using Rust's `From` & `Into`?

Reply:
1. Ability to work in no-std environments (don't rely on `std::convert::From`
2. Keep some flexibility as the API is still in flux - it may not remain exactly the same as `From`
3. To be explicit about the lifetime